Added GMRES solver

k7miller opened the pull request:
- Added GMRES iterative solver: GMRESSolver.hpp, GMRESSolver.cpp.
- Added testing code for GMRES: GMRESSolverTests.cpp.
- Added helper functions for basic linear algebra operations: LinAlgHelpers.hpp, LinAlgHelpers.cpp.

k7miller committed:
Added GMRES solver
- Added GMRES iterative solver: GMRESSolver.hpp, GMRESSolver.cpp.
- Added testing code for GMRES: GMRESSolverTests.cpp.
- Added helper functions for linear basic algebra operations: LinAlgHelpers.hpp, LinAlgHelpers.cpp.

k7miller committed:
Merge branch 'master' into gmresSolver

dsteinmo commented on include/GMRESSolver.hpp:
```diff
@@ -0,0 +1,352 @@
+#include <vector>
+
+namespace blitzdg {
+namespace details {
```
I think I'd like to see namespaces start a new level of indentation.

k7miller replied:
Sure thing bud.

dsteinmo commented on include/GMRESSolver.hpp:
```diff
@@ -0,0 +1,352 @@
+// unpreconditioned residual of the kth iterate.
+class GMRESSolver {
+public:
+    // Input/Output:
```
Nice documentation, can we change the commenting syntax to use the /**  */ style used elsewhere in the project? Otherwise doxygen won't know to pull these into the docs

k7miller replied:
Roger that Frank, I will have to review the comments in the code to see how this is done. I am not familiar with Doxygen, so I will have to check that out too.

dsteinmo commented on src/test/GMRESSolverTests.cpp:
```diff
@@ -0,0 +1,110 @@
+        bool operator()(const vector_type& in, vector_type& out) const {
+            blitz::firstIndex i;
+            blitz::secondIndex j;
+            out = blitz::sum(mref_(i, j) * in(j), j);
```
Super legit testing

dsteinmo commented on src/test/GMRESSolverTests.cpp:
```diff
@@ -0,0 +1,110 @@
+            GMRESSolver& gmres = *solver;
+            vector_type soln(N);
+            for (index_type i = 0; i < N; ++i)
+                soln(i) = rand() / (real_type(1) + RAND_MAX);
```
Why the random numbers? Is this for initial guess?

k7miller replied:
Yah, just wanted to test it with a nonzero initial guess.

k7miller replied:
I believe that blitz uses reference semantics for copy construction. So, the member variable Matvec::mat_ should be created as an alias of the input matrix. Because of this behavior, I don't think we need to worry about any overhead incurred by passing the Matvec functor by value. Clearly, this isn't an issue here, since A is only 5 x 5, but it would be good to know in general. There is also the reference() function in blitz, which may do the same thing.

dsteinmo commented on src/GMRESSolver.cpp:
```diff
@@ -0,0 +1,97 @@
+        throw invalid_argument("GMRESParams: divTol <= 0");
+}
+
+ostream& operator<<(ostream& strm, const GMRESOut& out) {
```
Very cool use of operator overloads

k7miller replied:
Thanks bud!

dsteinmo reviewed (approved):
Looks amazing!  I approve, but let's hold off on merging until we sort out the segmentation fault in Travis.

dsteinmo commented on src/test/GMRESSolverTests.cpp:
```diff
@@ -0,0 +1,110 @@
+    GMRESSolver* solver = nullptr;
+
+    Describe(GMRESSolver_Object) {
+        void Setup() {
```
I think you need to capitalize the U in SetUp, but maybe check the other test files. This could be why it breaks in the Linux build (accessing something that's not allocated yet)

k7miller committed:
Bug fix in GMRESSolverTests.cpp

dsteinmo committed:
Try not deleting the GMRES solver in TearDown?

dsteinmo committed:
Initialize some arrays where we assume zeros in some places that might not be.

dsteinmo committed:
valgrind: this should be a delete[] and not a straight-up delete.

dsteinmo committed:
More minor memory fixes from valgrind suggestions.

dsteinmo committed:
Get travis to stash the binaries to S3 so we can debug them.

dsteinmo committed:
Allow to deploy to S3 from this branch (not just master)...

dsteinmo commented:
Hey buddy, I hacked the build to stash the 'bin/test' binary somewhere and ran valgrind against it. It looks like blitz is having trouble figuring out the bounds of the input array in MatVec's operator that lives in the test class.

`==25705== Conditional jump or move depends on uninitialised value(s)
==25705==    at 0x425990: blitz::bounds::compute_ordering(int, int, int) (bounds.h:40)
==25705==    by 0x4D7B87: blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> >::ordering(int) const (expr.h:945)
==25705==    by 0x4D3F9B: blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >::ordering(int) const (expr.h:203)
==25705==    by 0x4D1C7E: blitz::_bz_ArrayExprReduce<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >, 1, blitz::ReduceSum<double, double> >::computeOrdering() (reduce.h:290)
==25705==    by 0x4CF88B: blitz::_bz_ArrayExprReduce<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >, 1, blitz::ReduceSum<double, double> >::_bz_ArrayExprReduce(blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >) (reduce.h:116)
==25705==    by 0x4CA8D6: blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprReduce<blitz::asExpr<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > > >::T_expr, 1, blitz::ReduceSum<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >::T_numtype, blitz::NumericTypeTraits<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >::T_numtype>::T_sumtype> > > blitz::sum<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >, 1>(blitz::ETBase<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > > > const&, blitz::IndexPlaceholder<1> const&) (in /home/dsteinmo/blitzdg/bin/test)
==25705==    by 0x4C72F1: GMRESSolverTests::Matvec::operator()(blitz::Array<double, 1> const&, blitz::Array<double, 1>&) const (GMRESSolverTests.cpp:21)
==25705==    by 0x4CB8B1: blitzdg::GMRESOut blitzdg::GMRESSolver::solve<GMRESSolverTests::Matvec, GMRESSolverTests::Precon>(GMRESSolverTests::Matvec, GMRESSolverTests::Precon, blitz::Array<double, 1> const&, blitz::Array<double, 1>&, blitzdg::GMRESParams const&) const (GMRESSolver.hpp:143)
==25705==    by 0x4C859F: GMRESSolverTests::GMRESSolver_Object::Solve_Ax_equals_b() (GMRESSolverTests.cpp:72)
==25705==    by 0x4E6F5A: igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::CallSpec(GMRESSolverTests::GMRESSolver_Object&, std::string const&, void (GMRESSolverTests::GMRESSolver_Object::*)(), igloo::TestResults&) (contextregistry.h:125)
==25705==    by 0x4E65B3: void igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::CallSpecs<GMRESSolverTests::GMRESSolver_Object>(std::map<std::string, igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::SpecInfo, std::less<std::string>, std::allocator<std::pair<std::string const, igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::SpecInfo> > > const&, std::string const&, igloo::TestResults&, igloo::TestListener&) (contextregistry.h:101)
==25705==    by 0x4E5E6E: void igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::Run<GMRESSolverTests::GMRESSolver_Object>(std::string const&, igloo::TestResults&, igloo::TestListener&) (contextregistry.h:73)`

The run time that's on travis might be being extra picky about something we're not catching locally, so it might be worth trying to refactor things so that the A matrix's memory lives on the operator (and not just a reference). I don't know exactly why this way isn't working though, just trying to see what's different here...

dsteinmo committed:
Revert "Allow to deploy to S3 from this branch (not just master)..."

This reverts commit 5cfa343b2d3c892c29c2a414e746d8798f7886ca.

dsteinmo committed:
Revert "Get travis to stash the binaries to S3 so we can debug them."

This reverts commit db1fc6a8cd15be5858a8af024a1cced61f06bab1.

dsteinmo committed:
Try changing the way we handle memory for gmres operators in tests.

dsteinmo commented:
Looks like Travis and AppVeyor agree with my changes. I'll let you look it over, and if you're cool with my refactor, let's ship it to master :shipit: .

k7miller committed:
Updated .gitignore file

k7miller committed:
Updated comments and formatting
- Indented namespace blocks.
- Updated comments to be consistent with Doxygen style.
- Added test to checkGMRESParams for the stagnation tolerance stgTol.
- Modified GMRESSolverTests.cpp to hopefully avoid seg fault error in linux builds.
- Modified function DROTG in LinAlgHelpers.hpp to explicitly set b = 0.

k7miller committed:
Merge branch 'gmresSolver' of https://github.com/WQCG/blitzdg into gmresSolver
- Resolved merge conflict by commenting out changes in GMRESSolverTests.cpp.

k7miller committed:
Fixed tests in GMRESSolverTests.cpp
